### PR TITLE
VZ-8616: ArgoCD multicluster fixes

### PR DESCRIPTION
**Pull request by leliyin**
VZ-8616: ArgoCD multicluster fixes

1. Use vmc name as cluster name to be shown in the argo UI
2. Check clusterroletemplatebinding exists before POST
3. Only create the token once and then on reconcile get it until it has an expiration

### CODE REVIEW

File: cluster-operator/controllers/vmc/argocd.go [hunk 1]
```diff
@@ -30,11 +31,13 @@ import (
...
-	createTimestamp                 = "verrazzano.io/create-timestamp"
-	expiresAtTimestamp              = "verrazzano.io/expires-at-timestamp"
-	clusterroletemplatebindingsPath = "/v3/clusterroletemplatebindings"
+	clusterSecretName                      = "argocd-cluster-secret"    //nolint:gosec
+	argocdClusterTokenTTLEnvVarName        = "ARGOCD_CLUSTER_TOKEN_TTL" //nolint:gosec
+	createTimestamp                        = "verrazzano.io/create-timestamp"
+	expiresAtTimestamp                     = "verrazzano.io/expires-at-timestamp"
+	clusterroletemplatebindingsPath        = "/v3/clusterroletemplatebindings"
+	clusterroletemplatebindingByUserIDPath = "/v3/clusterroletemplatebindings?userId="
+	dataField                              = "data"
```
Comment (fredtibbitts) [inline cluster-operator/controllers/vmc/argocd.go:40]: This constant is only used in one place, and it's very generic, I would just eliminate the const.

File: cluster-operator/controllers/vmc/argocd.go [hunk 2]
```diff
@@ -170,7 +173,7 @@ func (r *VerrazzanoManagedClusterReconciler) createOrUpdateArgoCDSecret(rc *ranc
 	return err
 }
 
-func (r *VerrazzanoManagedClusterReconciler) mutateArgoCDClusterSecret(secret *corev1.Secret, rc *rancherutil.RancherConfig, clusterID, cluserName string, rancherURL string, caData []byte) error {
+func (r *VerrazzanoManagedClusterReconciler) mutateArgoCDClusterSecret(secret *corev1.Secret, rc *rancherutil.RancherConfig, clusterName, clusterID, rancherURL string, caData []byte) error {
```
Comment (fredtibbitts) [inline cluster-operator/controllers/vmc/argocd.go:176]: This function is getting fairly large. It should be refactored into multiple smaller functions.

**Review by fredtibbitts — APPROVED** [review]
A couple of comments but they can be addressed in a separate PR. The logic seems OK to me.

File: cluster-operator/controllers/vmc/argocd.go [hunk 3]
```diff
@@ -249,6 +264,28 @@ func (r *VerrazzanoManagedClusterReconciler) updateArgoCDClusterRoleBindingTempl
...
+	// Send a request to see if the clusterroletemplatebinding for the given user exists
+	reqURL := rc.BaseURL + clusterroletemplatebindingByUserIDPath + url.PathEscape(userID) + "&clusterId=" + url.PathEscape(vmc.Status.RancherRegistration.ClusterID)
+	headers := map[string]string{"Authorization": "Bearer " + rc.APIAccessToken}
+	response, body, err := rancherutil.SendRequest(http.MethodGet, reqURL, headers, "", rc, r.log)
+	if err != nil {
+		return err
+	}
+	if response.StatusCode != http.StatusOK && response.StatusCode != http.StatusNotFound {
+		return err
+	}
```
Comment (davisgomes) [inline cluster-operator/controllers/vmc/argocd.go:276]: I'm not sure I understand this block. Won't err always be nil here?

File: cluster-operator/controllers/vmc/argocd.go [hunk 4]
```diff
@@ -249,6 +264,28 @@ func (r *VerrazzanoManagedClusterReconciler) updateArgoCDClusterRoleBindingTempl
...
 	}
 
+	// Send a request to see if the clusterroletemplatebinding for the given user exists
+	reqURL := rc.BaseURL + clusterroletemplatebindingByUserIDPath + url.PathEscape(userID) + "&clusterId=" + url.PathEscape(vmc.Status.RancherRegistration.ClusterID)
+	headers := map[string]string{"Authorization": "Bearer " + rc.APIAccessToken}
+	response, body, err := rancherutil.SendRequest(http.MethodGet, reqURL, headers, "", rc, r.log)
+	if err != nil {
+		return err
+	}
+	if response.StatusCode != http.StatusOK && response.StatusCode != http.StatusNotFound {
```
Comment (davisgomes) [inline cluster-operator/controllers/vmc/argocd.go:274]: We should not ignore status not found. This request should always return 200 even if the item does not exist

File: cluster-operator/controllers/vmc/argocd.go [hunk 5]
```diff
@@ -249,6 +264,28 @@ func (r *VerrazzanoManagedClusterReconciler) updateArgoCDClusterRoleBindingTempl
...
+		return err
+	}
+
+	if response.StatusCode == http.StatusOK {
```
Comment (davisgomes) [inline cluster-operator/controllers/vmc/argocd.go:278]: I think you can just check `if statusCode != http.StatusOK` then return a new error

Comment (davisgomes) [reply cluster-operator/controllers/vmc/argocd.go:278]: Then you can move the rest of the code in this block down outside of the if statement.

File: pkg/rancherutil/rancher_config.go [hunk 6]
```diff
@@ -233,32 +228,47 @@ func CreateTokenWithTTL(rc *RancherConfig, log vzlog.VerrazzanoLogger, ttl, clus
 		return "", "", log.ErrorfNewErr("Failed to parse response: %v", err)
 	}
 
-	return tokenResponse.Token, tokenResponse.Name, nil
+	return tokenResponse.Token, tokenResponse.Created, nil
+}
+
+type TokenAttrs struct {
```
Comment (davisgomes) [inline pkg/rancherutil/rancher_config.go:234]: Can you rename the two token structs to something more consistent i.e. `TokenPostResponse` and `TokenGetResponse`

File: cluster-operator/controllers/vmc/argocd.go (hunk 2 above)
Comment (leliyin) [reply cluster-operator/controllers/vmc/argocd.go:176]: will file a separate ticket to address this.

File: cluster-operator/controllers/vmc/argocd.go (hunk 1 above)
Comment (leliyin) [reply cluster-operator/controllers/vmc/argocd.go:40]: Done.

File: pkg/rancherutil/rancher_config.go (hunk 6 above)
Comment (leliyin) [reply pkg/rancherutil/rancher_config.go:234]: Fixed.

File: cluster-operator/controllers/vmc/argocd.go (hunk 5 above)
Comment (leliyin) [reply cluster-operator/controllers/vmc/argocd.go:278]: Fixed.

File: cluster-operator/controllers/vmc/argocd.go (hunk 3 above)
Comment (leliyin) [reply cluster-operator/controllers/vmc/argocd.go:276]: Fixed.

**Review by davisgomes — APPROVED** [review]

**Review by fredtibbitts — APPROVED** [review]

**Pull request merged**